diff --git a/etc/pgbouncer.ini b/etc/pgbouncer.ini
index 056ee0e..50d67f3 100644
--- a/etc/pgbouncer.ini
+++ b/etc/pgbouncer.ini
@@ -355,5 +355,12 @@ auth_file = /etc/pgbouncer/userlist.txt
 ;service_name = pgbouncer
 ;job_name = pgbouncer
 
+;; Disable support for prepared statements
+;; (P, B en C commands only together with prepareThreshold>0 in JDBC driver).
+;disable_prepared_statement_support = 1
+
+;; Number of prepared statements to cache on a backend connection.
+;prepared_statement_cache_queries = 100
+
 ;; Read additional config from other file
 ;%include /etc/pgbouncer/pgbouncer-other.ini
