diff --git a/include/bouncer.h b/include/bouncer.h
index f2c2bac..048ab8f 100644
--- a/include/bouncer.h
+++ b/include/bouncer.h
@@ -96,6 +96,7 @@ typedef union PgAddr PgAddr;
 typedef enum SocketState SocketState;
 typedef struct PktHdr PktHdr;
 typedef struct ScramState ScramState;
+typedef struct PgPreparedStatement PgPreparedStatement;
 
 extern int cf_sbuf_len;
 
@@ -118,6 +119,8 @@ extern int cf_sbuf_len;
 #include "janitor.h"
 #include "hba.h"
 #include "pam.h"
+#include "protocol_message.h"
+#include "prepared_statement.h"
 
 #ifndef WIN32
 #define DEFAULT_UNIX_SOCKET_DIR "/tmp"
@@ -254,6 +257,10 @@ struct PgStats {
 	usec_t xact_time;	/* total transaction time in us */
 	usec_t query_time;	/* total query time in us */
 	usec_t wait_time;	/* total time clients had to wait */
+
+	uint64_t ps_server_parse_count;
+	uint64_t ps_client_parse_count;
+	uint64_t ps_bind_count;
 };
 
 /*
@@ -490,6 +497,10 @@ struct PgDatabase {
 	struct AATree user_tree;	/* users that have been queried on this database */
 };
 
+struct OutstandingParsePacket {
+	struct List node;
+	bool ignore:1;
+};
 
 /*
  * A client or server connection.
@@ -567,6 +578,8 @@ struct PgSocket {
 
 	VarCache vars;		/* state of interesting server parameters */
 
+	struct PreparedStatementSocketState ps_state;		/* state of prepared statements on this socket */
+
 	SBuf sbuf;		/* stream buffer, must be last */
 };
 
@@ -683,6 +696,9 @@ extern char *cf_server_tls_cert_file;
 extern char *cf_server_tls_key_file;
 extern char *cf_server_tls_ciphers;
 
+extern int cf_disable_prepared_statement_support;
+extern int cf_prepared_statement_cache_queries;
+
 extern const struct CfLookup pool_mode_map[];
 
 extern usec_t g_suspend_start;
