diff --git a/include/objects.h b/include/objects.h
index 7f64ce1..f2af816 100644
--- a/include/objects.h
+++ b/include/objects.h
@@ -32,6 +32,10 @@ extern struct Slab *peer_pool_cache;
 extern struct Slab *pool_cache;
 extern struct Slab *user_cache;
 extern struct Slab *iobuf_cache;
+extern struct Slab *server_ps_cache;
+extern struct Slab *client_ps_cache;
+extern struct Slab *client_ps_query_cache;
+extern struct Slab *client_ps_param_cache;
 
 PgDatabase *find_peer(int peer_id);
 PgDatabase *find_database(const char *name);
