diff --git a/include/objects.h b/include/objects.h
index 33a0a3c..83b6096 100644
--- a/include/objects.h
+++ b/include/objects.h
@@ -28,6 +28,10 @@ extern struct Slab *db_cache;
 extern struct Slab *pool_cache;
 extern struct Slab *user_cache;
 extern struct Slab *iobuf_cache;
+extern struct Slab *server_ps_cache;
+extern struct Slab *client_ps_cache;
+extern struct Slab *client_ps_query_cache;
+extern struct Slab *client_ps_param_cache;
 
 PgDatabase *find_database(const char *name);
 PgUser *find_user(const char *name);
