diff --git a/src/admin.c b/src/admin.c
index 10b2a0b..15c0c69 100644
--- a/src/admin.c
+++ b/src/admin.c
@@ -588,8 +588,8 @@ static bool admin_show_users(PgSocket *admin, const char *arg)
 	return true;
 }
 
-#define SKF_STD "sssssisiTTiiississ"
-#define SKF_DBG "sssssisiTTiiississiiiiiii"
+#define SKF_STD "sssssisiTTiiississii"
+#define SKF_DBG "sssssisiTTiiississiiiiiiiii"
 
 static void socket_header(PktBuf *buf, bool debug)
 {
@@ -600,6 +600,7 @@ static void socket_header(PktBuf *buf, bool debug)
 				    "wait", "wait_us", "close_needed",
 				    "ptr", "link", "remote_pid", "tls",
 				    "application_name",
+					"ps_count", "ps_bytes",
 				    /* debug follows */
 				    "recv_pos", "pkt_pos", "pkt_remain",
 				    "send_pos", "send_remain",
@@ -663,6 +664,8 @@ static void socket_row(PktBuf *buf, PgSocket *sk, const char *state, bool debug)
 			     sk->close_needed,
 			     ptrbuf, linkbuf, remote_pid, infobuf,
 			     application_name ? application_name->str : "",
+			     cf_disable_prepared_statement_support ? 0 : is_server_socket(sk) ? HASH_COUNT(sk->server_prepared_statements) : HASH_COUNT(sk->prepared_statements),
+			     cf_disable_prepared_statement_support ? 0 : is_server_socket(sk) ? sizeof_server_ps_cache(sk) : sizeof_client_ps_cache(sk),
 			     /* debug */
 			     io ? io->recv_pos : 0,
 			     io ? io->parse_pos : 0,
