diff --git a/src/admin.c b/src/admin.c
index 0cf2e8b..abc4e92 100644
--- a/src/admin.c
+++ b/src/admin.c
@@ -500,10 +500,11 @@ static bool admin_show_databases(PgSocket *admin, const char *arg)
 		return true;
 	}
 
-	pktbuf_write_RowDescription(buf, "ssissiiisiiii",
+	pktbuf_write_RowDescription(buf, "ssissiiisiiiii",
 				    "name", "host", "port",
 				    "database", "force_user", "pool_size", "min_pool_size", "reserve_pool",
-				    "pool_mode", "max_connections", "current_connections", "paused", "disabled");
+				    "pool_mode", "ps_cache_queries", "max_connections", "current_connections",
+				    "paused", "disabled");
 	statlist_for_each(item, &database_list) {
 		db = container_of(item, PgDatabase, head);
 
@@ -512,13 +513,14 @@ static bool admin_show_databases(PgSocket *admin, const char *arg)
 		cv.value_p = &db->pool_mode;
 		if (db->pool_mode != POOL_INHERIT)
 			pool_mode_str = cf_get_lookup(&cv);
-		pktbuf_write_DataRow(buf, "ssissiiisiiii",
+		pktbuf_write_DataRow(buf, "ssissiiisiiiii",
 				     db->name, db->host, db->port,
 				     db->dbname, f_user,
 				     db->pool_size >= 0 ? db->pool_size : cf_default_pool_size,
 				     db->min_pool_size >= 0 ? db->min_pool_size : cf_min_pool_size,
 				     db->res_pool_size >= 0 ? db->res_pool_size : cf_res_pool_size,
 				     pool_mode_str,
+					 db->prepared_statement_cache_queries >= 0 ? db->prepared_statement_cache_queries : cf_prepared_statement_cache_queries,
 				     database_max_connections(db),
 				     db->connection_count,
 				     db->db_paused,
@@ -616,8 +618,8 @@ static bool admin_show_users(PgSocket *admin, const char *arg)
 	return true;
 }
 
-#define SKF_STD "sssssisiTTiiississ"
-#define SKF_DBG "sssssisiTTiiississiiiiiii"
+#define SKF_STD "sssssisiTTiiississii"
+#define SKF_DBG "sssssisiTTiiississiiiiiiiii"
 
 static void socket_header(PktBuf *buf, bool debug)
 {
@@ -848,6 +850,108 @@ static bool admin_show_active_sockets(PgSocket *admin, const char *arg)
 	return true;
 }
 
+#define PSF_STD "ssssTTiiiiii"
+
+static void socket_ps_header(PktBuf *buf)
+{
+	pktbuf_write_RowDescription(buf, PSF_STD,
+				"type", "user", "database", "state",
+				"connect_time", "request_time",
+				"cache_size", "cache_hit",
+				"req_parses", "exec_parses", "binds", "closes");
+}
+
+static void socket_ps_row(PktBuf *buf, PgSocket *sk, const char *state)
+{
+	int cache_size;
+
+	if (!is_prepared_statements_enabled(sk->pool)) {
+		return;
+	}
+
+	cache_size = is_server_socket(sk) ? \
+		HASH_COUNT(sk->ps_state.server_ps_entries) : HASH_COUNT(sk->ps_state.client_ps_entries);
+
+	pktbuf_write_DataRow(buf, PSF_STD,
+			     is_server_socket(sk) ? "S" :"C",
+			     sk->login_user ? sk->login_user->name : "(nouser)",
+			     sk->pool ? sk->pool->db->name : "(nodb)",
+			     state,
+			     sk->connect_time,
+			     sk->request_time,
+			     cache_size,
+			     sk->ps_state.binds_total > 0 ? (int) ((100*sk->ps_state.binds_total - sk->ps_state.executed_parses_total)/sk->ps_state.binds_total) : 0,
+			     sk->ps_state.requested_parses_total,
+			     sk->ps_state.executed_parses_total,
+			     sk->ps_state.binds_total,
+			     sk->ps_state.closes_total);
+}
+
+/* Helper for SHOW CLIENT_PS/SERVER_PS */
+static void show_ps_socket_list(PktBuf *buf, struct StatList *list, const char *state)
+{
+	struct List *item;
+	PgSocket *sk;
+
+	statlist_for_each(item, list) {
+		sk = container_of(item, PgSocket, head);
+		socket_ps_row(buf, sk, state);
+	}
+}
+
+/* Command: SHOW CLIENT PS */
+static bool admin_show_client_ps(PgSocket *admin, const char *arg)
+{
+	struct List *item;
+	PgPool *pool;
+	PktBuf *buf = pktbuf_dynamic(256);
+
+	if (!buf) {
+		admin_error(admin, "no mem");
+		return true;
+	}
+
+	socket_ps_header(buf);
+	statlist_for_each(item, &pool_list) {
+		pool = container_of(item, PgPool, head);
+		show_ps_socket_list(buf, &pool->active_client_list, "active");
+		show_ps_socket_list(buf, &pool->waiting_client_list, "waiting");
+		show_ps_socket_list(buf, &pool->active_cancel_req_list, "active_cancel_req");
+		show_ps_socket_list(buf, &pool->waiting_cancel_req_list, "waiting_cancel_req");
+	}
+
+	admin_flush(admin, buf, "SHOW");
+	return true;
+}
+
+/* Command: SHOW SERVER PS */
+static bool admin_show_server_ps(PgSocket *admin, const char *arg)
+{
+	struct List *item;
+	PgPool *pool;
+	PktBuf *buf = pktbuf_dynamic(256);
+
+	if (!buf) {
+		admin_error(admin, "no mem");
+		return true;
+	}
+
+	socket_ps_header(buf);
+	statlist_for_each(item, &pool_list) {
+		pool = container_of(item, PgPool, head);
+		show_ps_socket_list(buf, &pool->active_server_list, "active");
+		show_ps_socket_list(buf, &pool->idle_server_list, "idle");
+		show_ps_socket_list(buf, &pool->used_server_list, "used");
+		show_ps_socket_list(buf, &pool->tested_server_list, "tested");
+		show_ps_socket_list(buf, &pool->new_server_list, "new");
+		show_ps_socket_list(buf, &pool->active_cancel_server_list, "active_cancel");
+		show_ps_socket_list(buf, &pool->being_canceled_server_list, "being_canceled");
+	}
+
+	admin_flush(admin, buf, "SHOW");
+	return true;
+}
+
 /* Command: SHOW POOLS */
 static bool admin_show_pools(PgSocket *admin, const char *arg)
 {
@@ -1474,6 +1578,7 @@ static bool admin_show_totals(PgSocket *admin, const char *arg)
 
 static struct cmd_lookup show_map [] = {
 	{"clients", admin_show_clients},
+	{"client_ps", admin_show_client_ps},
 	{"config", admin_show_config},
 	{"databases", admin_show_databases},
 	{"fds", admin_show_fds},
@@ -1483,6 +1588,7 @@ static struct cmd_lookup show_map [] = {
 	{"peer_pools", admin_show_peer_pools},
 	{"pools", admin_show_pools},
 	{"servers", admin_show_servers},
+	{"server_ps", admin_show_server_ps},
 	{"sockets", admin_show_sockets},
 	{"active_sockets", admin_show_active_sockets},
 	{"stats", admin_show_stats},
