diff --git a/src/main.c b/src/main.c
index 3fc1a0c..ea173d7 100644
--- a/src/main.c
+++ b/src/main.c
@@ -187,6 +187,9 @@ char *cf_server_tls_cert_file;
 char *cf_server_tls_key_file;
 char *cf_server_tls_ciphers;
 
+int cf_disable_prepared_statement_support;
+int cf_prepared_statement_cache_queries;
+
 /*
  * config file description
  */
@@ -324,6 +327,9 @@ CF_ABS("user", CF_STR, cf_username, CF_NO_RELOAD, NULL),
 #endif
 CF_ABS("verbose", CF_INT, cf_verbose, 0, NULL),
 
+CF_ABS("disable_prepared_statement_support", CF_INT, cf_disable_prepared_statement_support, 0, "1"),
+CF_ABS("prepared_statement_cache_queries", CF_INT, cf_prepared_statement_cache_queries, 0, "100"),
+
 {NULL}
 };
 
@@ -927,10 +933,11 @@ int main(int argc, char *argv[])
 			break;
 		case 'V':
 			printf("%s\n", PACKAGE_STRING);
-			printf("libevent %s\nadns: %s\ntls: %s\n",
+			printf("libevent %s\nadns: %s\ntls: %s\nps: %s\n",
 			       event_get_version(),
 			       adns_get_backend(),
-			       tls_backend_version());
+			       tls_backend_version(),
+			       ps_version());
 #ifdef USE_SYSTEMD
 			printf("systemd: yes\n");
 #endif
@@ -1041,9 +1048,9 @@ int main(int argc, char *argv[])
 
 	write_pidfile();
 
-	log_info("process up: %s, libevent %s (%s), adns: %s, tls: %s", PACKAGE_STRING,
+	log_info("process up: %s, libevent %s (%s), adns: %s, tls: %s, ps: %s", PACKAGE_STRING,
 		 event_get_version(), event_base_get_method(pgb_event_base), adns_get_backend(),
-		 tls_backend_version());
+		 tls_backend_version(), ps_version());
 
 	sd_notify(0, "READY=1");
 
