diff --git a/src/main.c b/src/main.c
index 3fc1a0c..69f67ea 100644
--- a/src/main.c
+++ b/src/main.c
@@ -187,6 +187,8 @@ char *cf_server_tls_cert_file;
 char *cf_server_tls_key_file;
 char *cf_server_tls_ciphers;
 
+int cf_prepared_statement_cache_queries;
+
 /*
  * config file description
  */
@@ -274,6 +276,7 @@ CF_ABS("peer_id", CF_INT, cf_peer_id, 0, "0"),
 CF_ABS("pidfile", CF_STR, cf_pidfile, CF_NO_RELOAD, ""),
 CF_ABS("pkt_buf", CF_INT, cf_sbuf_len, CF_NO_RELOAD, "4096"),
 CF_ABS("pool_mode", CF_LOOKUP(pool_mode_map), cf_pool_mode, 0, "session"),
+CF_ABS("prepared_statement_cache_queries", CF_INT, cf_prepared_statement_cache_queries, 0, "0"),
 CF_ABS("query_timeout", CF_TIME_USEC, cf_query_timeout, 0, "0"),
 CF_ABS("query_wait_timeout", CF_TIME_USEC, cf_query_wait_timeout, 0, "120"),
 CF_ABS("cancel_wait_timeout", CF_TIME_USEC, cf_cancel_wait_timeout, 0, "10"),
@@ -927,10 +930,11 @@ int main(int argc, char *argv[])
 			break;
 		case 'V':
 			printf("%s\n", PACKAGE_STRING);
-			printf("libevent %s\nadns: %s\ntls: %s\n",
+			printf("libevent %s\nadns: %s\ntls: %s\nps: %s\n",
 			       event_get_version(),
 			       adns_get_backend(),
-			       tls_backend_version());
+			       tls_backend_version(),
+			       ps_version());
 #ifdef USE_SYSTEMD
 			printf("systemd: yes\n");
 #endif
@@ -1041,9 +1045,9 @@ int main(int argc, char *argv[])
 
 	write_pidfile();
 
-	log_info("process up: %s, libevent %s (%s), adns: %s, tls: %s", PACKAGE_STRING,
+	log_info("process up: %s, libevent %s (%s), adns: %s, tls: %s, ps: %s", PACKAGE_STRING,
 		 event_get_version(), event_base_get_method(pgb_event_base), adns_get_backend(),
-		 tls_backend_version());
+		 tls_backend_version(), ps_version());
 
 	sd_notify(0, "READY=1");
 
