diff --git a/src/stats.c b/src/stats.c
index 95d5cf4..921d011 100644
--- a/src/stats.c
+++ b/src/stats.c
@@ -30,6 +30,10 @@ static void reset_stats(PgStats *stat)
 	stat->xact_count = 0;
 	stat->xact_time = 0;
 	stat->wait_time = 0;
+
+  stat->ps_client_parse_count = 0;
+  stat->ps_server_parse_count = 0;
+  stat->ps_bind_count = 0;
 }
 
 static void stat_add(PgStats *total, PgStats *stat)
@@ -41,12 +45,19 @@ static void stat_add(PgStats *total, PgStats *stat)
 	total->xact_count += stat->xact_count;
 	total->xact_time += stat->xact_time;
 	total->wait_time += stat->wait_time;
+
+  total->ps_client_parse_count += stat->ps_client_parse_count;
+  total->ps_server_parse_count += stat->ps_server_parse_count;
+  total->ps_bind_count += stat->ps_bind_count;
 }
 
 static void calc_average(PgStats *avg, PgStats *cur, PgStats *old)
 {
 	uint64_t query_count;
 	uint64_t xact_count;
+  uint64_t ps_client_parse_count;
+  uint64_t ps_server_parse_count;
+  uint64_t ps_bind_count;
 
 	usec_t dur = get_cached_time() - old_stamp;
 
@@ -71,6 +82,14 @@ static void calc_average(PgStats *avg, PgStats *cur, PgStats *old)
 		avg->xact_time = (cur->xact_time - old->xact_time) / xact_count;
 
 	avg->wait_time = USEC * (cur->wait_time - old->wait_time) / dur;
+
+  ps_client_parse_count = cur->ps_client_parse_count - old->ps_client_parse_count;
+  ps_server_parse_count = cur->ps_server_parse_count - old->ps_server_parse_count;
+  ps_bind_count = cur->ps_bind_count - old->ps_bind_count;
+
+  avg->ps_client_parse_count = USEC * ps_client_parse_count / dur;
+  avg->ps_server_parse_count = USEC * ps_server_parse_count / dur;
+  avg->ps_bind_count = USEC * ps_bind_count / dur;
 }
 
 static void write_stats(PktBuf *buf, PgStats *stat, PgStats *old, char *dbname)
@@ -369,6 +388,18 @@ static void refresh_stats(evutil_socket_t s, short flags, void *arg)
 			 avg.client_bytes, avg.server_bytes,
 			 avg.xact_time, avg.query_time,
 			 avg.wait_time);
+    
+    if (!cf_disable_prepared_statement_support) {
+      log_info("prepared statement stats: %" PRIu64 " client parses/s (total %" PRIu64 "),"
+          " %" PRIu64 " server parses/s (total %" PRIu64 "),"
+          " %" PRIu64 " binds/s (total %" PRIu64 ")",
+          avg.ps_client_parse_count,
+          cur_total.ps_client_parse_count,
+          avg.ps_server_parse_count,
+          cur_total.ps_server_parse_count,
+          avg.ps_bind_count,
+          cur_total.ps_bind_count);
+    }
 	}
 
 	sd_notifyf(0,
